# Google Merchandise EDA

#### Importing the Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

#### Ignoring the warnings

In [2]:
import warnings
warnings.filterwarnings("ignore")

#### Loading the Dataset

In [3]:
events=pd.read_csv('data/events1.csv')
items=pd.read_csv('data/items.csv')
users=pd.read_csv('data/users.csv')

#### Merging the 3 Data Frame into one complete Data Frame

In [4]:
items = items.rename(columns={'id': 'item_id'})
users = users.rename(columns={'id': 'user_id'})

df = (
    events
    .merge(items, on='item_id', how='left')
    .merge(users, on='user_id', how='left')
)

#### Displaying the first 10 rows of the dataset to get a quick look at the data

In [5]:
df.head(10)

user_id  ga_session_id country   device         type  item_id  \
0     2133          16909      US   mobile     purchase       94   
1     2133          16909      US   mobile     purchase      425   
2     5789          16908      SE  desktop     purchase        1   
3     5789          16908      SE  desktop     purchase       62   
4     5808           4267      US   mobile  add_to_cart      842   
5     5808           4267      US   mobile  add_to_cart      951   
6     5808           4267      US   mobile  add_to_cart      950   
7     5808           4267      US   mobile  add_to_cart     1068   
8     5808           4267      US   mobile  add_to_cart      862   
9     5808           4267      US   mobile  add_to_cart     1119   

                date_x                                       name    brand  \
0  2020-11-01 00:27:14                    Google Large Tote White   Google   
1  2020-11-01 00:27:14          Google Heather Green Speckled Tee   Google   
2  2020-11-01 01:44:44                             Google KeepCup   Google   
3  2020-11-01 01:44:44                      Google Mini Kick Ball   Google   
4  2020-11-01 03:06:29             Google Infant Hero Onesie Grey   Google   
5  2020-11-01 03:06:29                      YouTube Icon Tee Grey  YouTube   
6  2020-11-01 03:06:29                   Google Navy Speckled Tee   Google   
7  2020-11-01 03:06:29                       Google Eco Tee Black   Google   
8  2020-11-01 03:06:29  Google Land & Sea French Terry Sweatshirt   Google   
9  2020-11-01 03:06:29                          Google Crew Socks   Google   

              variant       category  price_in_usd  ltv               date_y  
0  Single Option Only           Bags            10   34  2020-11-01 00:09:55  
1                  XL        Apparel            21   34  2020-11-01 00:09:55  
2  Single Option Only            New            28   72  2020-10-14 00:26:03  
3  Single Option Only            Fun             2   72  2020-10-14 00:26:03  
4                 NaN        Apparel            25    0  2020-08-18 03:44:52  
5                 NaN        Apparel            22    0  2020-08-18 03:44:52  
6                 NaN        Apparel            30    0  2020-08-18 03:44:52  
7                 NaN        Apparel            22    0  2020-08-18 03:44:52  
8                 NaN  Shop by Brand            55    0  2020-08-18 03:44:52  
9                 NaN        Apparel            16    0  2020-08-18 03:44:52

#### Displaying the total number of rows and columns in the dataset

In [6]:
rows, columns = df.shape
print(f'This complete contains {rows} rows and {columns} columns.')

This complete contains 758884 rows and 14 columns.


#### Displaying column names

In [7]:
df.columns

Index(['user_id', 'ga_session_id', 'country', 'device', 'type', 'item_id',
       'date_x', 'name', 'brand', 'variant', 'category', 'price_in_usd', 'ltv',
       'date_y'],
      dtype='str')

#### Based on the publicly available Google Merchandise Store dataset, the meaning of each column is as follows:

- user_id: Unique identifier assigned to each user/customer.
- ga_session_id: Unique identifier for a user session (a single visit to the website).
- country: Country from which the user accessed the store.
- device: Type of device used during the session (e.g., mobile, desktop).
- type: Type of event performed (e.g., purchase, add_to_cart).
- item_id: Unique identifier of the product involved in the event.
- date_x: Timestamp of the event (when the interaction such as purchase or add_to_cart occurred).
- name: Name of the product.
- brand: Brand associated with the product (e.g., Google, YouTube).
- variant: Specific variation of the product (e.g., size, color, or option). May contain missing values.
- category: Product category (e.g., Apparel, Bags, Fun, New).
- price_in_usd: Price of the product in US dollars.
- ltv: Lifetime Value of the user, total revenue generated by the user over time.
- date_y: Timestamp related to the user’s lifecycle (often first interaction or account creation date).

In [8]:
df.describe()

user_id  ga_session_id        item_id   price_in_usd  \
count  758884.000000  758884.000000  758884.000000  758884.000000   
mean    29541.809665    3632.325492     885.629356      24.759137   
std     55129.337846    3952.875337     289.267126      20.892459   
min         1.000000       0.000000       0.000000       1.000000   
25%      8401.000000     625.000000     913.000000      11.000000   
50%     17181.000000    2103.000000     972.000000      22.000000   
75%     25809.000000    5365.000000    1040.000000      30.000000   
max    270145.000000   18033.000000    1380.000000     313.000000   

                 ltv  
count  758884.000000  
mean       61.428210  
std       124.049917  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%        80.000000  
max      1530.000000

#### Displaying columns along with their data types

In [9]:
df.dtypes

user_id          int64
ga_session_id    int64
country            str
device             str
type               str
item_id          int64
date_x             str
name               str
brand              str
variant            str
category           str
price_in_usd     int64
ltv              int64
date_y             str
dtype: object

#### Observations on Data Types

Upon inspecting the dataset, most columns have appropriate data types for analysis. However, a few columns may need attention or conversion:

- country, device, type, name, brand, variant, category, date_x, date_y – Stored as str (string). For analysis, date_x and date_y should ideally be converted to datetime objects.

- user_id, ga_session_id, item_id, price_in_usd, ltv – Stored as numeric types (int64) and ready for analysis.

#### Checking all unique values of each column

In [10]:
for col in df.columns:
    unique_vals = df[col].unique()
    print(f'Column Name : {col}')
    print(f'Total number of unique values : {len(unique_vals)}')
    print(f'Unique Values : {unique_vals}')
    print('-'*50)
    print('\n')

Column Name : user_id
Total number of unique values : 14701
Unique Values : [  2133   5789   5808 ...  19742 266597 234237]
--------------------------------------------------


Column Name : ga_session_id
Total number of unique values : 18034
Unique Values : [16909 16908  4267 ...  9347  5568 12443]
--------------------------------------------------


Column Name : country
Total number of unique values : 109
Unique Values : <StringArray>
['US', 'SE', 'CA', 'BE', 'GB', 'TR', 'PE', 'ES', 'IN', 'MY',
 ...
 'GH', 'OM', 'BY', 'BA', 'PY', 'PA', 'XK', 'TT', 'MM', 'AM']
Length: 109, dtype: str
--------------------------------------------------


Column Name : device
Total number of unique values : 3
Unique Values : <StringArray>
['mobile', 'desktop', 'tablet']
Length: 3, dtype: str
--------------------------------------------------


Column Name : type
Total number of unique values : 3
Unique Values : <StringArray>
['purchase', 'add_to_cart', 'begin_checkout']
Length: 3, dtype: str
-----------

## Summary of Data Observations

Columns with correct types
- user_id, ga_session_id, item_id, price_in_usd, ltv – Stored as int64, ready for analysis.
- country, device, type, name, brand, variant, category, date_x, date_y – Stored as strings; mostly fine, though date_x and date_y should be converted to datetime for time-based analysis

Columns needing cleaning or further preprocessing
- variant – Contains missing values (NaN) and extra spaces in strings; may need stripping or filling.
- date_x and date_y – Currently strings; convert to datetime for session or purchase analysis over time.
- device and type – Few unique values; can be converted to categorical type for memory efficiency.

Columns with anomalies or special observations
- price_in_usd – Ranges from $1 to $313; ensure consistency and check for outliers.
- ltv – Ranges from 0 to 1530; some users have zero lifetime value (likely no purchases).
- variant – Some entries are nonspecific like "No options available" or "Single Option Only".
- category – 21 unique values, includes special categories like "Gift Cards" and "Black Lives Matter"

Columns likely not needed for feature analysis
- user_id, ga_session_id, and item_id are identifiers, useful for merging or aggregation, but may not be directly used as features in modeling.

#### Checking all unique values of each column

In [11]:
# Count of NaN / None values per column
nan_count = df.isnull().sum()

# Count of empty strings per column
empty_count = (df == '').sum()

# Total missing values (NaN + empty strings)
total_missing = nan_count + empty_count

# Percentage of missing values
missing_percent = (total_missing / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Values': total_missing,
    'Percentage (%)': missing_percent
}).sort_values(by='Percentage (%)', ascending=False)

print(missing_df)

               Missing Values  Percentage (%)
variant                636260       83.841536
country                  4555        0.600223
ga_session_id               0        0.000000
device                      0        0.000000
type                        0        0.000000
user_id                     0        0.000000
item_id                     0        0.000000
date_x                      0        0.000000
name                        0        0.000000
brand                       0        0.000000
category                    0        0.000000
price_in_usd                0        0.000000
ltv                         0        0.000000
date_y                      0        0.000000


#### Data Cleaning

- variant has 83.84% missing values, which too high to impute reliably.
- country has only 0.6% missing values, so a small fraction that can be safely filled with 'Unknown' so as to avoid any biases.

In [12]:
# Fill missing values in the 'country' column with 'Unknown'
df['country'].fillna('Unknown', inplace=True)

0         US
1         US
2         SE
3         SE
4         US
          ..
758879    IN
758880    IN
758881    IN
758882    IN
758883    IN
Name: country, Length: 758884, dtype: str

In [13]:
# Columns you want to remove
columns_to_remove = ['variant'] 

# Remove columns
df = df.drop(columns=columns_to_remove)

In [14]:
# Convert date columns to datetime
df['date_x'] = pd.to_datetime(df['date_x'])
df['date_y'] = pd.to_datetime(df['date_y'])

# Convert categorical columns to 'category' dtype
categorical_cols = ['country', 'device', 'type', 'name', 'brand', 'category']
for col in categorical_cols:
    df[col] = df[col].astype('category')

# Ensure identifiers and numeric columns stay as int64
numeric_cols = ['user_id', 'ga_session_id', 'item_id', 'price_in_usd', 'ltv']
for col in numeric_cols:
    df[col] = df[col].astype('int64')

print(df.dtypes)

user_id                   int64
ga_session_id             int64
country                category
device                 category
type                   category
item_id                   int64
date_x           datetime64[us]
name                   category
brand                  category
category               category
price_in_usd              int64
ltv                       int64
date_y           datetime64[us]
dtype: object


## Time for Insights

The graphs created in the analysis below had to be commented. Reasob being that - if we upload the files with the graphs itelf, the file size exceeds above 100 MB and Git hub does not allow such big files to be uploaded

#### Univaraite Analysis

In [15]:
'''# -------------------------
# CATEGORICAL VARIABLES
# -------------------------

# 1. Top 10 Countries
top_countries = df['country'].value_counts().nlargest(10).rename_axis('Country').reset_index(name='Count')
fig = px.bar(top_countries, x='Count', y='Country', orientation='h',
             title='Top 10 Countries by Users', text='Count')
fig.update_traces(textposition='outside')
fig.update_layout(yaxis=dict(categoryorder='total ascending'))
fig.show()

# 2. Device Distribution
device_counts = df['device'].value_counts().rename_axis('Device').reset_index(name='Count')
fig = px.pie(device_counts, names='Device', values='Count', title='Device Distribution')
fig.show()

# 3. Event Type Distribution
type_counts = df['type'].value_counts().rename_axis('Event Type').reset_index(name='Count')
fig = px.bar(type_counts, x='Event Type', y='Count', text='Count', title='Event Type Distribution')
fig.update_traces(textposition='outside')
fig.show()

# 4. Top 10 Categories
top_categories = df['category'].value_counts().nlargest(10).rename_axis('Category').reset_index(name='Count')
fig = px.bar(top_categories, x='Category', y='Count', text='Count', title='Top 10 Categories')
fig.update_traces(textposition='outside')
fig.update_xaxes(tickangle=45)
fig.show()

# 5. Brand Distribution
brand_counts = df['brand'].value_counts().rename_axis('Brand').reset_index(name='Count')
fig = px.pie(brand_counts, names='Brand', values='Count', title='Brand Distribution')
fig.show()'''

"# -------------------------\n# CATEGORICAL VARIABLES\n# -------------------------\n\n# 1. Top 10 Countries\ntop_countries = df['country'].value_counts().nlargest(10).rename_axis('Country').reset_index(name='Count')\nfig = px.bar(top_countries, x='Count', y='Country', orientation='h',\n             title='Top 10 Countries by Users', text='Count')\nfig.update_traces(textposition='outside')\nfig.update_layout(yaxis=dict(categoryorder='total ascending'))\nfig.show()\n\n# 2. Device Distribution\ndevice_counts = df['device'].value_counts().rename_axis('Device').reset_index(name='Count')\nfig = px.pie(device_counts, names='Device', values='Count', title='Device Distribution')\nfig.show()\n\n# 3. Event Type Distribution\ntype_counts = df['type'].value_counts().rename_axis('Event Type').reset_index(name='Count')\nfig = px.bar(type_counts, x='Event Type', y='Count', text='Count', title='Event Type Distribution')\nfig.update_traces(textposition='outside')\nfig.show()\n\n# 4. Top 10 Categories\nt

In [16]:
'''# -------------------------
# NUMERICAL VARIABLES
# -------------------------

# 6. Price Distribution
fig = px.histogram(df, x='price_in_usd', nbins=30, marginal='box', 
                   title='Price Distribution', text_auto=True)
fig.show()

# 7. Lifetime Value (LTV) Distribution
fig = px.histogram(df, x='ltv', nbins=30, marginal='box',
                   title='Customer LTV Distribution', text_auto=True)
fig.show()

# 8. Top 10 Items Sold
top_items = df['item_id'].value_counts().nlargest(10).rename_axis('Item ID').reset_index(name='Count')
fig = px.bar(top_items, x='Item ID', y='Count', text='Count', title='Top 10 Items Sold')
fig.update_traces(textposition='outside')
fig.show()

# 9. Sessions per User
sessions_per_user = df.groupby('user_id')['ga_session_id'].nunique().reset_index(name='Sessions')
fig = px.histogram(sessions_per_user, x='Sessions', nbins=30, title='Sessions per User', text_auto=True)
fig.show()'''

"# -------------------------\n# NUMERICAL VARIABLES\n# -------------------------\n\n# 6. Price Distribution\nfig = px.histogram(df, x='price_in_usd', nbins=30, marginal='box', \n                   title='Price Distribution', text_auto=True)\nfig.show()\n\n# 7. Lifetime Value (LTV) Distribution\nfig = px.histogram(df, x='ltv', nbins=30, marginal='box',\n                   title='Customer LTV Distribution', text_auto=True)\nfig.show()\n\n# 8. Top 10 Items Sold\ntop_items = df['item_id'].value_counts().nlargest(10).rename_axis('Item ID').reset_index(name='Count')\nfig = px.bar(top_items, x='Item ID', y='Count', text='Count', title='Top 10 Items Sold')\nfig.update_traces(textposition='outside')\nfig.show()\n\n# 9. Sessions per User\nsessions_per_user = df.groupby('user_id')['ga_session_id'].nunique().reset_index(name='Sessions')\nfig = px.histogram(sessions_per_user, x='Sessions', nbins=30, title='Sessions per User', text_auto=True)\nfig.show()"

In [17]:
'''# -------------------------
# DATE / TIME VARIABLES
# -------------------------

# 10. Purchase Timeline
fig = px.histogram(df, x='date_x', nbins=50, title='Purchase Activity Over Time')
fig.show()'''

"# -------------------------\n# DATE / TIME VARIABLES\n# -------------------------\n\n# 10. Purchase Timeline\nfig = px.histogram(df, x='date_x', nbins=50, title='Purchase Activity Over Time')\nfig.show()"

#### Bivaraite Analysis

In [18]:
df['revenue'] = df['price_in_usd']  # for revenue-based analysis

In [19]:
'''# Numeric vs Numeric
# 1. Price vs LTV
fig = px.scatter(df, x='price_in_usd', y='ltv', color='category', hover_data=['name'], title='Price vs LTV')
fig.show()

# 2. Items per session vs LTV
items_per_session = df.groupby(['user_id','ga_session_id']).size().reset_index(name='Num_Items')
user_ltv = df.groupby('user_id')['ltv'].max().reset_index()
merged = items_per_session.merge(user_ltv, on='user_id')
fig = px.scatter(merged, x='Num_Items', y='ltv', title='Number of Items per Session vs LTV')
fig.show()'''


"# Numeric vs Numeric\n# 1. Price vs LTV\nfig = px.scatter(df, x='price_in_usd', y='ltv', color='category', hover_data=['name'], title='Price vs LTV')\nfig.show()\n\n# 2. Items per session vs LTV\nitems_per_session = df.groupby(['user_id','ga_session_id']).size().reset_index(name='Num_Items')\nuser_ltv = df.groupby('user_id')['ltv'].max().reset_index()\nmerged = items_per_session.merge(user_ltv, on='user_id')\nfig = px.scatter(merged, x='Num_Items', y='ltv', title='Number of Items per Session vs LTV')\nfig.show()"

In [20]:
'''# Numeric vs Categorical (averaged)
# 3. Average Price by Category (Top 10)
top_categories_list = top_categories['Category'].tolist()
df_top_cat = df[df['category'].isin(top_categories_list)]
avg_price_cat = df_top_cat.groupby('category')['price_in_usd'].mean().reset_index()
fig = px.bar(avg_price_cat, x='category', y='price_in_usd', text='price_in_usd', title='Average Price by Category (Top 10)')
fig.update_traces(textposition='outside')
fig.update_xaxes(tickangle=45)
fig.show()

# 4. Average LTV by Device
avg_ltv_device = df.groupby('device')['ltv'].mean().reset_index()
fig = px.bar(avg_ltv_device, x='device', y='ltv', text='ltv', title='Average LTV by Device')
fig.update_traces(textposition='outside')
fig.show()

# 5. Average LTV by Country (Top 10)
df_top_country = df[df['country'].isin(top_countries['Country'].tolist())]
avg_ltv_country = df_top_country.groupby('country')['ltv'].mean().reset_index()
fig = px.bar(avg_ltv_country, x='country', y='ltv', text='ltv', title='Average LTV by Country (Top 10)')
fig.update_traces(textposition='outside')
fig.update_xaxes(tickangle=45)
fig.show()'''

"# Numeric vs Categorical (averaged)\n# 3. Average Price by Category (Top 10)\ntop_categories_list = top_categories['Category'].tolist()\ndf_top_cat = df[df['category'].isin(top_categories_list)]\navg_price_cat = df_top_cat.groupby('category')['price_in_usd'].mean().reset_index()\nfig = px.bar(avg_price_cat, x='category', y='price_in_usd', text='price_in_usd', title='Average Price by Category (Top 10)')\nfig.update_traces(textposition='outside')\nfig.update_xaxes(tickangle=45)\nfig.show()\n\n# 4. Average LTV by Device\navg_ltv_device = df.groupby('device')['ltv'].mean().reset_index()\nfig = px.bar(avg_ltv_device, x='device', y='ltv', text='ltv', title='Average LTV by Device')\nfig.update_traces(textposition='outside')\nfig.show()\n\n# 5. Average LTV by Country (Top 10)\ndf_top_country = df[df['country'].isin(top_countries['Country'].tolist())]\navg_ltv_country = df_top_country.groupby('country')['ltv'].mean().reset_index()\nfig = px.bar(avg_ltv_country, x='country', y='ltv', text='ltv'

In [21]:
'''# Categorical vs Categorical
# 6. Event Type vs Device
event_device = df.groupby(['type','device']).size().reset_index(name='Count')
fig = px.bar(event_device, x='type', y='Count', color='device', text='Count', title='Event Type by Device')
fig.update_traces(textposition='outside')
fig.show()

# 7. Category vs Brand (Top 10 categories)
cat_brand = df[df['category'].isin(top_categories_list)].groupby(['category','brand']).size().reset_index(name='Count')
fig = px.bar(cat_brand, x='category', y='Count', color='brand', text='Count', title='Category vs Brand (Top 10 Categories)')
fig.update_traces(textposition='outside')
fig.update_xaxes(tickangle=45)
fig.show()

# Date vs Numeric
# 8. Total Revenue over Time
daily_revenue = df.groupby('date_x')['revenue'].sum().reset_index()
fig = px.line(daily_revenue, x='date_x', y='revenue', title='Total Revenue Over Time')
fig.show()

# 9. Number of Purchases per Day
daily_purchases = df.groupby('date_x').size().reset_index(name='Num_Purchases')
fig = px.line(daily_purchases, x='date_x', y='Num_Purchases', title='Number of Purchases per Day')
fig.show()'''

"# Categorical vs Categorical\n# 6. Event Type vs Device\nevent_device = df.groupby(['type','device']).size().reset_index(name='Count')\nfig = px.bar(event_device, x='type', y='Count', color='device', text='Count', title='Event Type by Device')\nfig.update_traces(textposition='outside')\nfig.show()\n\n# 7. Category vs Brand (Top 10 categories)\ncat_brand = df[df['category'].isin(top_categories_list)].groupby(['category','brand']).size().reset_index(name='Count')\nfig = px.bar(cat_brand, x='category', y='Count', color='brand', text='Count', title='Category vs Brand (Top 10 Categories)')\nfig.update_traces(textposition='outside')\nfig.update_xaxes(tickangle=45)\nfig.show()\n\n# Date vs Numeric\n# 8. Total Revenue over Time\ndaily_revenue = df.groupby('date_x')['revenue'].sum().reset_index()\nfig = px.line(daily_revenue, x='date_x', y='revenue', title='Total Revenue Over Time')\nfig.show()\n\n# 9. Number of Purchases per Day\ndaily_purchases = df.groupby('date_x').size().reset_index(name=

#### Multivaraite Analysis

In [22]:
'''# -------------------------
# MULTIVARIATE ANALYSIS
# -------------------------

# 1. Device × Category × Average LTV (Top 10 categories)
df_top_cat = df[df['category'].isin(top_categories_list)]
avg_ltv_device_cat = df_top_cat.groupby(['category','device'])['ltv'].mean().reset_index()
fig = px.bar(avg_ltv_device_cat, x='category', y='ltv', color='device', text='ltv', title='Average LTV by Device and Category (Top 10)')
fig.update_traces(textposition='outside')
fig.update_xaxes(tickangle=45)
fig.show()

# Top 5 countries & top 5 categories average price
top_5_countries = top_countries['Country'].head(5).tolist()
top_5_categories = top_categories['Category'].head(5).tolist()

df_filtered = df[(df['country'].isin(top_5_countries)) & 
                 (df['category'].isin(top_5_categories))]

avg_price_country_cat = df_filtered.groupby(['country','category'])['price_in_usd'].mean().reset_index()

fig = px.bar(
    avg_price_country_cat,
    x='country',
    y='price_in_usd',
    color='category',
    text='price_in_usd',
    title='Average Price by Country and Category (Top 5 Countries & Categories)',
    barmode='group'
)
fig.update_traces(textposition='outside')
fig.update_xaxes(tickangle=45)
fig.show()

# 3. Event Type × Device × Total LTV
type_device_ltv = df.groupby(['type','device']).agg({'ltv':'sum'}).reset_index()
fig = px.bar(type_device_ltv, x='type', y='ltv', color='device', text='ltv', title='Total LTV by Event Type and Device')
fig.update_traces(textposition='outside')
fig.show()

# 4. Category × LTV × Date (Top 5 categories)
# Top 5 categories for LTV over time
top_5_categories = top_categories['Category'].head(5).tolist()

# Filter data for Top 5 categories
cat_date_ltv_top5 = df[df['category'].isin(top_5_categories)].groupby(['date_x','category'])['ltv'].sum().reset_index()

# Plot LTV over time by category (Top 5)
fig = px.line(
    cat_date_ltv_top5,
    x='date_x',
    y='ltv',
    color='category',
    title='LTV Over Time by Category (Top 5)'
)
fig.show()'''

"# -------------------------\n# MULTIVARIATE ANALYSIS\n# -------------------------\n\n# 1. Device × Category × Average LTV (Top 10 categories)\ndf_top_cat = df[df['category'].isin(top_categories_list)]\navg_ltv_device_cat = df_top_cat.groupby(['category','device'])['ltv'].mean().reset_index()\nfig = px.bar(avg_ltv_device_cat, x='category', y='ltv', color='device', text='ltv', title='Average LTV by Device and Category (Top 10)')\nfig.update_traces(textposition='outside')\nfig.update_xaxes(tickangle=45)\nfig.show()\n\n# Top 5 countries & top 5 categories average price\ntop_5_countries = top_countries['Country'].head(5).tolist()\ntop_5_categories = top_categories['Category'].head(5).tolist()\n\ndf_filtered = df[(df['country'].isin(top_5_countries)) & \n                 (df['category'].isin(top_5_categories))]\n\navg_price_country_cat = df_filtered.groupby(['country','category'])['price_in_usd'].mean().reset_index()\n\nfig = px.bar(\n    avg_price_country_cat,\n    x='country',\n    y='pri